### PR TITLE
*: Introduce runaway statement in resource group

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #43691

Problem Summary:

### What is changed and how it works?
After https://github.com/pingcap/kvproto/pull/1122, replace glog goleak function name

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by CabinfeverB** [comment]
cc @glorv @nolouch @Connor1996

File: parser/parser_test.go [hunk 1]
```diff
@@ -3700,6 +3701,16 @@ func TestDDL(t *testing.T) {
 		{"create resource group x ru_per_sec=3000 burstable", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 3000 BURSTABLE"},
 		{"create resource group x burstable ru_per_sec=4000", true, "CREATE RESOURCE GROUP `x` BURSTABLE RU_PER_SEC = 4000"},
 		{"create resource group x ru_per_sec=20, priority=LOW, burstable", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 20 PRIORITY = LOW BURSTABLE"},
+		{"create resource group x ru_per_sec=1000 QUERY LIMIT EXEC_ELAPSED_IN_SEC '10s' ACTION DRYRUN", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 1000 QUERY LIMIT EXEC_ELAPSED_IN_SEC = '10s' ACTION DRYRUN"},
```
Comment (nolouch) [inline parser/parser_test.go:3704]: IMO, I would like to make all runaway spec into one option, it looks more layered. like:
```
create resource group x ru_per_sec=1000 RUNAWAY_QUERY='{ LIMIT_EXEC_ELAPSED_IN_SEC="10s", ACTION=KILL}'
```

File: parser/parser.y [hunk 2]
```diff
@@ -1725,6 +1737,87 @@ ResourceGroupPriorityOption:
...
+			yylex.AppendError(yylex.Errorf("The EXEC_ELAPSED_IN_SEC option is not a valid duration: %s", err.Error()))
+			return 1
+		}
+		$$ = &ast.ResourceGroupRunawayOption{Tp: ast.RunawayRule, StrValue: $5}
+	}
+|	"ACTION" EqOpt ResourceGroupRunawayActionOption
+	{
+		$$ = &ast.ResourceGroupRunawayOption{Tp: ast.RunawayAction, IntValue: $3.(int32)}
+	}
+|	"WATCH" ResourceGroupRunawayWatchOption "DURATION" EqOpt stringLit
```
Comment (Connor1996) [inline parser/parser.y:1811]: let's put `EqOpt` after "WATCH"

**Review by Connor1996 — APPROVED** [review]
rest LGTM

**Comment by Connor1996** [comment]
PTAL again @nolouch

**Review by hawkingrei — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

File: parser/ast/ddl.go [hunk 3]
```diff
@@ -2156,8 +2158,62 @@ func (n *ResourceGroupOption) Restore(ctx *format.RestoreCtx) error {
 			ctx.WriteString(n.StrValue)
 		case ResourceBurstableOpiton:
 			ctx.WriteKeyWord("BURSTABLE")
+		case ResourceGroupRunaway:
+			if len(n.ResourceGroupRunawayOptionList) > 0 {
+				ctx.WritePlain("QUERY_LIMIT=(")
```
Comment (glorv) [inline parser/ast/ddl.go:2163]:
```suggestion
				ctx.WriteKeyWord("QUERY_LIMIT")
				ctx.WritePlain(" = (")
```

File: parser/model/model.go [hunk 4]
```diff
@@ -1849,20 +1849,85 @@ func (p *PlacementSettings) Clone() *PlacementSettings {
...
+type RunawayWatchType int32
+
+//revive:disable:exported
+const (
+	WatchExact RunawayWatchType = iota
+	WatchSimilar
+)
+
+// RunawayWatchValueToName converts the runaway watch value to corresponding name
+func RunawayWatchValueToName(value int32) string {
```
Comment (glorv) [inline parser/model/model.go:1873]:
```suggestion
func (t RunawayWatchType) String() string {
```

File: parser/parser_test.go [hunk 5]
```diff
@@ -3710,6 +3711,19 @@ func TestDDL(t *testing.T) {
 		{"create resource group x ru_per_sec=3000 burstable", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 3000 BURSTABLE"},
 		{"create resource group x burstable ru_per_sec=4000", true, "CREATE RESOURCE GROUP `x` BURSTABLE RU_PER_SEC = 4000"},
 		{"create resource group x ru_per_sec=20, priority=LOW, burstable", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 20 PRIORITY = LOW BURSTABLE"},
+		{"create resource group x ru_per_sec=1000 QUERY_LIMIT=(EXEC_ELAPSED_IN_SEC '10s' ACTION DRYRUN)", true, "CREATE RESOURCE GROUP `x` RU_PER_SEC = 1000 QUERY_LIMIT=(EXEC_ELAPSED_IN_SEC = '10s' ACTION DRYRUN)"},
```
Comment (glorv) [inline parser/parser_test.go:3714]: The output format is not consistent with each other, `(EXEC_ELAPSED_IN_SEC = '10s' ACTION DRYRUN)` should be `(EXEC_ELAPSED_IN_SEC = '10s' ACTION = DRYRUN)`

File: parser/ast/ddl.go [hunk 6]
```diff
@@ -2156,8 +2158,62 @@ func (n *ResourceGroupOption) Restore(ctx *format.RestoreCtx) error {
...
+			ctx.WriteString(n.StrValue)
+		case RunawayAction:
+			ctx.WriteKeyWord("ACTION ")
+			ctx.WriteKeyWord(model.RunawayActionValueToName(n.IntValue))
+		case RunawayWatch:
+			ctx.WriteKeyWord("WATCH ")
+			ctx.WriteKeyWord(model.RunawayWatchValueToName(n.IntValue))
+			ctx.WritePlain(" ")
+			ctx.WriteKeyWord("DURATION ")
+			ctx.WritePlain("= ")
```
Comment (Connor1996) [inline parser/ast/ddl.go:2213]: should after `WATCH` instead of `DURATION`

**Review by glorv — APPROVED** [review]

**Pull request merged**